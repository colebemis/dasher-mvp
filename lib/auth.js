import cookies from 'js-cookie'
import { getGithubToken } from '../lib/github'

export const GITHUB_TOKEN_KEY = 'github_token'

export async function logIn({ githubCode, res }) {
  const githubToken = await getGithubToken(githubCode)

  res.setHeader(
    'Set-Cookie',
    `${GITHUB_TOKEN_KEY}=${githubToken}; SameSite=Strict`,
  )
}

export function logOut() {
  cookies.remove(GITHUB_TOKEN_KEY)
}
