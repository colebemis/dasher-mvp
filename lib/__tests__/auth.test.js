import cookies from 'js-cookie'
import { GITHUB_TOKEN_KEY, logIn, logOut, isLoggedIn } from '../auth'

jest.mock('../../lib/github')

describe('logIn', () => {
  it('sets appropriate cookies', async () => {
    const args = {
      githubCode: 'TEST_GITHUB_CODE',
      res: {
        setHeader: jest.fn(),
      },
    }

    await logIn(args)

    expect(args.res.setHeader.mock.calls).toMatchSnapshot()
  })
})

describe('logOut', () => {
  it('removes appropriate cookies', () => {
    cookies.set(GITHUB_TOKEN_KEY, 'TEST_GITHUB_TOKEN')

    expect(cookies.get(GITHUB_TOKEN_KEY)).toBe('TEST_GITHUB_TOKEN')

    logOut()

    expect(cookies.get(GITHUB_TOKEN_KEY)).toBeNull()
  })
})

describe('isLoggedIn', () => {
  it('returns true when logged in server-side', () => {
    const req = {
      headers: {
        cookie: `${GITHUB_TOKEN_KEY}=TEST_GITHUB_TOKEN`,
      },
    }

    expect(isLoggedIn(req)).toBe(true)
  })

  it('returns false when logged out server-side', () => {
    const req = {
      headers: {
        cookie: '',
      },
    }

    expect(isLoggedIn(req)).toBe(false)
  })

  it('returns true when logged in client-side', () => {
    cookies.set(GITHUB_TOKEN_KEY, 'TEST_GITHUB_TOKEN')

    expect(isLoggedIn()).toBe(true)

    cookies.remove(GITHUB_TOKEN_KEY)
  })

  it('returns false when logged out client-side', () => {
    expect(isLoggedIn()).toBe(false)
  })
})
