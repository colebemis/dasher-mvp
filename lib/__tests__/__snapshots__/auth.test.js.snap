// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logIn sets appropriate cookies 1`] = `
Array [
  Array [
    "github_token",
    "TEST_GITHUB_TOKEN",
  ],
]
`;
